### ISSUE TITLE
Double SG values at same timestamp

### ISSUE THREAD

**Issue by mortlind** [pazaan/decoding-contour-next-link#28]
Once in a while, I get double SG values for some period of time when extracting sensor events; this may be from one sample to some 5-10 samples. The attached plot from my web application shows this behaviour. Any clue? I do admit that I have not thoroughly traced the origin of these extra data, but I will do that. 

My best theory about the reason for these branches is the pump losing contact with the sensor transmitter for a short time. The transmitter stores the SG-values, and once contact with the pump is established, the pump can read out the saved SG values. These values supposedly have an offset time with respect to some other point. In the current code this offset time is added to the time of the reference point, whereas they should have been subtracted.

This is becoming an increasing problem, since we observe that the frequency of the pump losing signal from the transmitter is increasing, and the durations longer. It can get quite difficult to grasp the  plot sometimes.

![20171126111236](https://user-images.githubusercontent.com/5364433/33239315-821bd1ee-d29f-11e7-81e5-7761a0541ea3.jpg)

P.S.: This issue was originally posted in the @paulokow repository.

**Comment by mortlind** [comment]
I believe the solution must lie in the handling of timestamps for nested sensor readings events at https://github.com/pazaan/decoding-contour-next-link/blob/b0198b721958230dc38a7899209b7de637f83c03/pump_history_parser.py#L399

My hypothesis is, that for a period where the pump could not get signal from the sensor tranmitter, and for which the data generated will be later read out, a separate `SensorGlucoseReadingsEvent` will be generated. The timestamp for this, aggregate, event will be at the end of the dead period of communication. The timestamps of the nested events have to be offset back in time, rather than, for a regular `SensorGlucoseReadingsEvent`, forward in time from the aggregate event timestamp.

If this is so, how to distinguish the two different types of `SensorGlucoseReadingsEvent`. I will try to investigate this.

**Comment by Pogman** [comment]
@mortlind hey, we already handle this as part of the android uploader. Here's the latest code for this:

```
        // validate that this contains a new SGV record
        if (pumpRecord.isCgmActive()) {
            if (pumpRecord.getEventRTC() - pumpRecord.getCgmRTC() > (POLL_PERIOD_MS + (POLL_GRACE_PERIOD_MS / 2)) / 1000)
                pumpRecord.setOldSgvWhenNewExpected(true);
            else if (!pumpRecord.isCgmWarmUp() && pumpRecord.getSgv() > 0 &&
                    activePump.getPumpHistory().where().equalTo("cgmRTC", pumpRecord.getCgmRTC()).findAll().size() == 0)
                pumpRecord.setValidSGV(true);
        }
```

Sometimes the pump will get a SGV late and it will be available this poll period and so a retry is done after 60-90 secs.

**Comment by mortlind** [comment]
@Pogman Thanks, but where do you have the code. Could you paste a reference? I do not find the same code or interface here, e.g. the method `.isCgmActive` is not to be found. If that method should be applicable in the current code base, it must be some property which can be parsed from the payload belonging to a `SensorGlucoseReadingsEvent` object. Rigth?

**Comment by Pogman** [comment]
That code is not online, probably will have a new branch up in a few days. 

When we poll we get the pump time and the the pump status. The pump time RTC is used as the RTC for this event and if the difference from the event and the CGM RTC is too big we've gotten an old SGV from the pump.

**Comment by Pogman** [comment]
Code from master that handles the old sgv from pump: 

https://github.com/pazaan/600SeriesAndroidUploader/blob/d582d467fa67d9062a1227e3e2ba8b77068f4d83/app/src/main/java/info/nightscout/android/medtronic/service/MedtronicCnlIntentService.java#L303

**Comment by mortlind** [comment]
@Pogman It seems that you are talking about extracting status? The symptom that I get arises when extracting sensor history data from the pump. Do you think that we are talking about two different issues here?

**Comment by Pogman** [comment]
Ha maybe... ;)

BTW I've worked extensively with history and it's always precise

**Comment by paulokow** [comment]
Hi @mortlind, can you tell if the issue is visible in the [dump you provided some time ago](https://github.com/pazaan/decoding-contour-next-link/blob/master/testdata/mortlind_20170923_cgm_sample.dat) 

Otherwise I'd need a new dump to be able to help on the issue (or wait 3 weeks, my son gets CGM provisioned in mid Dec).

It may be that I ported something badly from Javascript back then :/

**Comment by mortlind** [comment]
@paulokow Thanks, I will make a new dump, and send it here. I will take a look at it too :-)

**Comment by mortlind** [comment]
@paulokow I looked a bit into the code now. If I dump data, the `allNestedEvents` method of `SensorGlucoseReadingsEvent`-objects have already been called, and the `SensorGlucoseReadingsEvent`-objects will be missing from the dumped data, replaced by the wrongly timestamped `SensorGlucoseReadingEvent`-objects. So we will have to modify the code to inspect the `SensorGlucoseReadingsEvent`-objects. But in the current code, the part that is parsed on the `eventData` for the  `SensorGlucoseReadingsEvent`-objects, there is no clue to distinguishing the direction of time offsets. We will have to inspect the 600SeriesAndroidUploader-project for that. Agreed?

**Comment by Pogman** [comment]
Hey - the 600Series history handling code is not online yet but maybe this will be of some use. It's just the code I used to make sure I was seeing the right output:
```
            } else if (eventType == EventType.SENSOR_GLUCOSE_READINGS_EXTENDED) {
                int minutesBetweenReadings = read8toUInt(eventData, index + 0x0B);
                int numberOfReadings = read8toUInt(eventData, index + 0x0C);
                int predictedSg = read16BEtoUInt(eventData, index + 0x0D);
                result += " Min:" + minutesBetweenReadings + " Num:" + numberOfReadings + " SGP:" + predictedSg;

                int pos = index + 15;
                for (int i = 0; i < numberOfReadings; i++) {

                    Date sgtimestamp = MessageUtils.decodeDateTime(eventRTC & 0xFFFFFFFFL - (i * minutesBetweenReadings * 60), eventOFFSET);

                    int sg = read16BEtoUInt(eventData, pos) & 0x03FF;
                    double isig = read16BEtoUInt(eventData, pos + 2) / 100.0;
                    double vctr = (((eventData[pos] >> 2 & 3) << 8) | eventData[pos + 4] & 0x000000FF) / 100.0;
                    double rateOfChange = read16BEtoInt(eventData, pos + 5) / 100.0;
                    int sensorStatus = read8toUInt(eventData, pos + 7);
                    int readingStatus = read8toUInt(eventData, pos + 8);

                    boolean backfilledData = (readingStatus & 1) == 1;
                    boolean settingsChanged = (readingStatus & 2) == 2;
                    boolean noisyData = sensorStatus == 1;
                    boolean discardData = sensorStatus == 2;
                    boolean sensorError = sensorStatus == 3;

                    int sensorException = 0;
                    int sensorEx = 0;

                    if (sg > 0x1FF) {
                        sensorException = sg & 0x00FF;
                        sg = 0;
                        result += "\n! " + sgtimestamp;
                    } else {
                        result += "\n* " + sgtimestamp;
                    }

                    result += " SGV:" + sg + " ISIG:" + isig + " VCTR:" + vctr + " ROC:" + rateOfChange + " STAT:" + readingStatus
                            + " BF:" + backfilledData + " SC:" + settingsChanged + " NS:" + noisyData + " DD:" + discardData + " SE:" + sensorError + " Exception:" + sensorException + "/" + sensorEx;

                    pos += 9;
                }
```

**Comment by mortlind** [comment]
@Pogman OK. I thought you meant that it was not distributed yet. I have looked for the snippet you sent in the 600SeriesAndroidUploader-project. But you mean that it is not in the repo? I see there a very different timestamp computation. I will look into the present code, and see if the necessary resources are in place to make the same computation here. Thanks!

**Comment by Pogman** [comment]
Sorry for the confusion. The 600series code that's in the repo only handles the time and status messages from the pump. The history handling and the posted snippet are not yet in the online repo.

**Comment by mortlind** [comment]
@Pogman Can you post more of the code? Specifically I need to be able to find out how `eventRTC` and `eventOFFSET` are computed. I suppose `eventRTC` is the same as in the present code called `self.timestamp` at https://github.com/pazaan/decoding-contour-next-link/blob/b0198b721958230dc38a7899209b7de637f83c03/pump_history_parser.py#L399

**Comment by Pogman** [comment]
code is up on my repo now: https://github.com/Pogman/640gAndroidUploader/tree/dev-history

**Comment by mortlind** [comment]
@Pogman Thanks. I will try to understand the code and time computation.

**Comment by paulokow** [comment]
Fixed in b7cd28b
The problem was with my porting. The single events are stored in the reverse order and one has to deduct time for each event (what actually makes a lot of sense, as it's impossible to see the events in the future!).
My code was adding the events forward, that's why these were overlapping with following events.

**Comment by mortlind** [comment]
@paulokow I have not investigated thoroughly, but I was searching for a distinction between `SensorGlucoseReadingsEvent`which where collected during connectivity and after a period of disconnectivity. I interpret your comment as follows: Under normal conditions, the single sensor events are contained in a `SensorGlucoseReadingsEvent` alone, i.e. cardinality 1. For these single readings recorded during connection between the pump and transmitter, there is no difference in the order they are traversed and since the index * timebetweenmeasurement is always zero, it does not matter. When the pump has missed events for a period, it gets these missed events from the transmitter in a burst inside a `SensorGlucoseReadingsEvent`, and now you corrected the arithmetics for calculating the timestamps on the series of missed events. Is that correctly understood?
